File: src/lambda-calculus-typed/church-numbers.ts
import { church_false, church_true } from './church-booleans';
import { first, newPair, Pair } from './pair';

// Church encoded natural number type definition
export type ChurchNumber = (f: (x: any) => any) => (x: any) => any;

// Arithmetic operations on Church encoded natural numbers
export const succ = (n: ChurchNumber) => (f: (x: any) => any) => (x: any) => f(n(f)(x)) as ChurchNumber;
export const sum = (n: ChurchNumber) => (m: ChurchNumber) => (f: (x: any) => any) => (x: any) =>
    n(succ(m))(f)(x) as ChurchNumber;
export const mult = (n: ChurchNumber) => (m: ChurchNumber) => (f: (x: any) => any) => (x: any) =>
    n(m(f))(x) as ChurchNumber;
export const power = (n: ChurchNumber) => (m: ChurchNumber) => (f: (x: any) => any) => (x: any) =>
    n(m)(f)(x) as ChurchNumber;

// Sutraction implemented using pairs
const nextPair = (pair: Pair<ChurchNumber, ChurchNumber>) => pair((_x) => (y) => newPair(y)(succ(y)));
export const pred = (n: ChurchNumber) => n(nextPair)(newPair(church_0)(church_0))(first) as ChurchNumber;
export const sub = (n: ChurchNumber) => (m: ChurchNumber) => m(pred)(n);

const always_false = ((x) => (_y: any) => x)(church_false);
export const isZero = (n: ChurchNumber) => n(always_false)(church_true);

// first 10 natural numbers
export const church_0: ChurchNumber = (_f) => (x) => x;
export const church_1 = succ(church_0);
export const church_2 = succ(church_1);
export const church_3 = succ(church_2);
export const church_4 = succ(church_3);
export const church_5 = succ(church_4);
export const church_6 = succ(church_5);
export const church_7 = succ(church_6);
export const church_8 = succ(church_7);
export const church_9 = succ(church_8);

// numberJs converts a church encoded number to a javascript number
export const numberJs = (n: ChurchNumber) => n((x) => x + 1)(0);

File: go/src/church-numbers.go
package lambdacalculus

var f Lambda = func(x interface{}) interface{} {
	return x.(int) + 1
}
var x = 0

// ChurchNumber are Church encoded numbers
type ChurchNumber = func(f interface{}) Lambda

// Zero is 0
var Zero ChurchNumber = func(f interface{}) Lambda {
	return func(x interface{}) interface{} {
		return x
	}
}

// Succ is the successor function
var Succ = func(n ChurchNumber) ChurchNumber {
	return func(f interface{}) Lambda {
		return func(x interface{}) interface{} {
			switch f := f.(type) {
			case func(ChurchNumber) ChurchNumber:
				xChurch := x.(ChurchNumber)
				return f(n(f)(xChurch).(ChurchNumber))
			case ChurchNumber:
				l := x.(Lambda)
				return f(n(f)(l))
			case func(Pair) Pair:
				xp := x.(Pair)
				return f(n(f)(xp).(Pair))
			default:
				return f.(Lambda)(n(f)(x))
			}
		}
	}
}

// Sum is the sum function
var Sum = func(n ChurchNumber) nm {
	return func(m ChurchNumber) ChurchNumber {
		return m(Succ)(n).(ChurchNumber)
	}
}

// Mult is the multiplication function
var Mult = func(n ChurchNumber) nm {
	return func(m ChurchNumber) ChurchNumber {
		return func(f interface{}) Lambda {
			return n(m(f))
		}
	}
}

// Pow is the power function
var Pow = func(n ChurchNumber) nxl {
	return func(m ChurchNumber) xl {
		return func(f interface{}) Lambda {
			return n(m)(f).(Lambda)
		}
	}
}

// IsZero returns true if the number is zero, false otherwise
var IsZero = func(n ChurchNumber) ChurchBoolean {
	return n(Constant(F))(T).(ChurchBoolean)
}

// nextPair returns a new pair where the first parameter is the second parameter of the original pair
// and the second parameter is the successor of the second parameter of the original pair
var nextPair = func(p Pair) Pair {
	return Tuple2Struct(p(Second))(Succ(p(Second).(ChurchNumber)))
}

// Prev is the previous number
var Prev = func(n ChurchNumber) ChurchNumber {
	p00 := Tuple2Struct(Zero)(Zero)
	return n(nextPair)(p00).(Pair)(First).(ChurchNumber)
}

// Sub subtracts m from n
var Sub = func(n ChurchNumber) nm {
	return func(m ChurchNumber) ChurchNumber {
		return m(Prev)(n).(ChurchNumber)
	}
}
